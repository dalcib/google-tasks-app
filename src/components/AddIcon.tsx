import * as React from 'react';

import ContentAdd from 'material-ui/svg-icons/content/add';

const AddIcon = ContentAdd;

export { AddIcon }