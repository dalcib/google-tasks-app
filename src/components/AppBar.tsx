import * as React from 'react';
import MuiAppBar from 'material-ui/AppBar';

interface AppBarProps {
    onMenuClick: () => void;
}

const AppBar = (props: AppBarProps) => (
    <MuiAppBar 
        className="app-bar"
        title="Tasks App"
        showMenuIconButton={true}
        onLeftIconButtonClick={props.onMenuClick}
    />
);

export { AppBar }