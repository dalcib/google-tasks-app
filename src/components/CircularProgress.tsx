import * as React from 'react';

import MuiCircularProgress from 'material-ui/CircularProgress';

const style: React.CSSProperties = {
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: -20,
    marginLeft: -20,
};

const CircularProgress = () => (
    <MuiCircularProgress style={style} />
);

export { CircularProgress }