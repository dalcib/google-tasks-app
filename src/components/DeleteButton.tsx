import * as React from 'react';

import { IconButton } from 'material-ui';

import { DeleteIcon } from './DeleteIcon';

interface DeleteButtonProps {
    onClick: (event: React.MouseEvent<any>) => void;
}

const DeleteButton = (props: DeleteButtonProps) => (
    <IconButton
        onClick={props.onClick}
    >
        <DeleteIcon />
    </IconButton>
);

export { DeleteButton, DeleteButtonProps }