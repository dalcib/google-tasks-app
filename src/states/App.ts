import { DataState } from "./Data";
import { TaskList } from "../models/TaskList";
import { Task } from "../models/Task";

class AppState {
    taskLists: DataState<TaskList>;
    tasks: DataState<Task>;
}

export { AppState }