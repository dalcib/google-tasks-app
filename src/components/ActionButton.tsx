import * as React from 'react';

import FloatingActionButton from 'material-ui/FloatingActionButton';
import { AddIcon } from './AddIcon';

const ActionButton = () => (
    <FloatingActionButton
        className="action-button"
    >
        <AddIcon />
    </FloatingActionButton>
);

export { ActionButton }