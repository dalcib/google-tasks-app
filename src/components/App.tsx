import * as React from 'react';

import { AppBar } from './AppBar';
import { LinearProgress } from './LinearProgress';
import { Menu } from './Menu';
import { TaskListViewContainer } from '../containers/TaskListView';
import { ActionButton } from './ActionButton';
import { TaskList } from '../models/TaskList';
import { Task } from '../models/Task';

interface AppProps {
    loading?: boolean;
    taskLists?: TaskList[];
    lastTaskId?: string;
    onStart?: () => void;
    onSelectList?: (taskListId: string) => void;
    onAddTask?: (taskListId: string, previousTaskId: string) => void;
}

interface AppState {
    openMenu: boolean;
    taskListId: string;
}

class App extends React.Component<AppProps, AppState> {

    componentWillMount(): void {
        this.start();
    }

    componentWillReceiveProps(nextProps: AppProps): void {
        const {
            taskLists,
            lastTaskId,
            loading
        } = nextProps;

        const notLoading = loading === false;
        const hasTaskLists = taskLists.length;
        const hasntSelectedList = this.state.taskListId === null;

        if (
            notLoading
            && hasTaskLists
            && hasntSelectedList
        ) {
            this.selectList(taskLists[0].id);
        }
    }

    render(): React.ReactElement<any> | false {
        const {
            loading,
            taskLists,
            onSelectList,
            onAddTask,
        } = this.props;

        const {
            openMenu,
            taskListId
        } = this.state;

        if (loading) {
            return (
                <section className="app">
                    <LinearProgress />
                </section>
            );
        }

        return (
            <section className="app">
                <AppBar
                    onMenuClick={() => this.toggleMenu()}
                />
                <Menu
                    taskLists={taskLists}
                    open={openMenu}
                    onCloseMenu={() => this.toggleMenu()}
                    onSelectList={taskListId => this.selectList(taskListId)}
                />
                <TaskListViewContainer taskListId={taskListId} />
                <ActionButton onClick={() => this.addTask()} />
            </section>
        );
    }

    private start(): void {
        const { onStart } = this.props;
        this.setState({ openMenu: false, taskListId: null });
        onStart();
    }

    private toggleMenu(): void {
        this.setState({ openMenu: !this.state.openMenu });
    }

    private selectList(taskListId: string) {
        const { onSelectList } = this.props;
        this.setState({ taskListId });
        onSelectList(taskListId);
    }

    private addTask(): void {
        const { lastTaskId, onAddTask } = this.props;
        const { taskListId } = this.state;
        onAddTask(taskListId, lastTaskId);
    }

}


export { App, AppProps }