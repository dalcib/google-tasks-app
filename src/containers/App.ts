import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { TaskActions } from '../actions/Task';
import { TaskListActions } from '../actions/TaskList';
import { App, AppProps } from '../components/App';
import { AppState } from '../states/App';

function mapStateToProps(appState: AppState): AppProps {
    const taskLists = appState.taskLists.data;
    const loading = appState.taskLists.fetching;
    const tasks = appState.tasks.data;

    const lastTask = tasks && tasks.length > 0 ?
        tasks[tasks.length - 1] :
        {};

    return {
        loading,
        taskLists,
        lastTaskId: lastTask.id
    };
}

function mapDispatchToProps(dispatch: Dispatch<any>): AppProps {
    return {
        onStart(): void {
            const action = TaskListActions.getTaskLists();
            dispatch(action);
        },
        onSelectList(taskListId: string): void {
            const action = TaskActions.getTasks(taskListId);
            dispatch(action);
        },
        onAddTask(taskListId: string, previousTaskId: string): void {
            const action = TaskActions.createTask(taskListId, previousTaskId);
            dispatch(action);
        }
    };
}

const AppContainer = connect(
    mapStateToProps,
    mapDispatchToProps
)(App as any);

export { AppContainer }