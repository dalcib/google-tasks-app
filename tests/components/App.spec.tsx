import * as React from 'react';
import * as renderer  from 'react-test-renderer';

import { App } from '../../src/components/App';

describe('Given an App', () => {
    
    beforeEach(() => {
        this.component = renderer.create(<App />);
        this.tree = this.component.toJSON();
    });

    it('should match snapshot', () => {
        expect(this.tree).toMatchSnapshot();
    });

});