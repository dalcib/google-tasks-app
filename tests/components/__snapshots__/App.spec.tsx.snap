// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an App should match snapshot 1`] = `
<h1>
  Hello World!
</h1>
`;
